src/main/java/org/broadinstitute/hellbender/tools/exome/sexgenotyper/PloidyAnnotatedTargetCollection.java

```diff
@@ -0,0 +1,219 @@
+package org.broadinstitute.hellbender.tools.exome.sexgenotyper;
+
+import com.google.common.collect.Sets;
+import htsjdk.samtools.util.Locatable;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.tools.exome.Target;
+import org.broadinstitute.hellbender.tools.exome.TargetCollection;
+import org.broadinstitute.hellbender.utils.IndexRange;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+
+import javax.annotation.Nonnull;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * A collection of {@link Target} instances along with helper methods for generating their genotype ploidy
+ * annotations on-the-fly using provided contig ploidy annotations.
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+public final class PloidyAnnotatedTargetCollection implements TargetCollection<Target> {
+
+    /**
+     * Map from targets to their ploidy annotations (based on target contigs)
+     */
+    private Map<Target, ContigPloidyAnnotation> targetToContigPloidyAnnotationMap;
```

Comment:
It is not necessary, but is is faster to do so; by design, `PloidyAnnotatedTargetCollection` is meant to act as a fast oracle for Target -> ContigPloidyAnnotations. It is worthwhile to do some preprocessing in the beginning for faster `getTargetPloidyByTag` calls.